### PR TITLE
Rust 1.39 Update + Spring Cleaning

### PR DESCRIPTION
**Pull request by dhrgit**
## Reason for This PR

Tidy is good, tidy is love.

## Description of Changes

- workspace: all sources are now placed in their respective crates, under `/src/<crate>`;
- workspace: unified the jailer source under a single bin crate `/src/jailer`; removed jailer dependency from the firecracker crate;
- workspace: configured the build artifacts dir via `.cargo/config`;
- devctr: removed env variable used to override the cargo build dir; updated to Rust toolchain 1.39;
- clippy: added safety doc comments;
- vsock: switched to using zeroed memory for the TX buffer, instead of `mem::uninitialized`, since the latter is both unsafe and deprecated by Rust 1.39;
- also fixed up some whitespace errors.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided. 
- [ ] The description of changes is clear and encompassing.
- [ ] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [ ] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [ ] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [ ] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.
- [ ] Either no new `unsafe` code has been added, or, the newly added `unsafe`
      code is unavoidable and properly documented.

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
Changes LGTM. I'll approve this once the arm build is fixed as well.

**Comment by acatangiu** [comment]
Do you have a quick reference resource you can link that explains why moving the crates under `/src` is tidy while the current state is not?

**Comment by andreeaflorescu** [comment]
My personal take on this is I that it makes more sense as the source code is clearly separated from the documentation and other utilities. It is much easier to navigate through misc docs and tools as they don't mingle with rust code as well.

**Comment by dhrgit** [comment]
> Do you have a quick reference resource you can link that explains why moving the crates under `/src` is tidy while the current state is not?

I don't have one handy. However, you can look at it like this: just as a crate is structured using a `src` dir for the sources, a `target` dir for the build artifacts, and the root for support files (such as `Cargo.toml`), so should a workspace be. I.e. sources go under `src`, integration tests go under `tests`, documentation goes under `docs`, and so on.

**Comment by acatangiu** [comment]
Makes sense. I thought we enable some `Cargo` magic as well (that we were previously missing).

File: src/devices/src/virtio/vsock/csm/txbuf.rs [hunk 1]
```diff
@@ -50,17 +49,9 @@ impl TxBuf {
             return Err(Error::TxBufFull);
         }
 
-        // We're using a closure here to return the boxed slice, instead of a value (i.e.
```
Comment (acatangiu) [inline src/devices/src/virtio/vsock/csm/txbuf.rs]: nit: Why did you remove this comment?

Comment (dhrgit) [reply src/devices/src/virtio/vsock/csm/txbuf.rs]: It was outdated. Clippy got fixed in the meantime, and we'd removed the clippy bypass, but forgot to also removed the comment.

**Comment by raduweiss** [comment]
My _esthetics only_ opinion is "+100".

File: src/utils/src/structs.rs [hunk 2]
```diff
@@ -41,13 +44,16 @@ pub unsafe fn read_struct<T: Copy, F: Read>(f: &mut F, out: &mut T) -> Result<()
...
 /// # Arguments
 ///
 /// * `f` - The input to read from.  Often this is a file.
 /// * `len` - The number of structs to fill with data read from `f`.
+///
+/// # Safety
+///
+/// This is unsafe because the structs are initialized to unverified data read from the input.
+/// `read_struct_slice` should only be called for plain old data structs.  It is not endian safe.
 // This lint check is now deprecated - https://github.com/rust-lang/rust-clippy/pull/3478/files
```
Comment (dianpopa) [inline src/utils/src/structs.rs]: Is this comment still valid? What lint check is it refferring to?

File: tests/integration_tests/build/test_unittests.py [hunk 3]
```diff
@@ -39,9 +39,6 @@ def test_unittests(test_session_root_path, target):
     if MACHINE == "x86_64":
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply tests/integration_tests/build/test_unittests.py]: A miss indeed. Fixed.

File: src/utils/src/structs.rs (hunk 2 above)
Comment (dhrgit) [reply src/utils/src/structs.rs]: Removed. Looks like that comment had been left behind when the clippy override was removed.

File: Cargo.toml [hunk 4]
```diff
@@ -1,30 +1,9 @@
...
-mmds = { path = "mmds" }
-seccomp = { path = "seccomp" }
-utils = { path = "utils" }
-vmm = { path = "vmm" }
-libc = ">=0.2.39"
-
-
-[dev-dependencies]
-tempfile = ">=3.0.2"
+[workspace]
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply Cargo.toml]: FIxed.
